packages/@aws-cdk/core/lib/deployment-configuration.ts

```diff
@@ -0,0 +1,635 @@
+import * as asset_schema from '@aws-cdk/cdk-assets-schema';
+import * as cxschema from '@aws-cdk/cloud-assembly-schema';
+import * as cxapi from '@aws-cdk/cx-api';
+import * as crypto from 'crypto';
+import * as fs from 'fs';
+import * as path from 'path';
+import { DockerImageAssetLocation, DockerImageAssetSource, FileAssetLocation, FileAssetPackaging, FileAssetSource } from './assets';
+import { Fn } from './cfn-fn';
+import { Construct, ConstructNode, IConstruct, ISynthesisSession } from './construct-compat';
+import { FileAssetParameters } from './private/asset-parameters';
+import { Stack } from './stack';
+import { Token } from './token';
+
+/**
+ * The well-known name for the docker image asset ECR repository. All docker
+ * image assets will be pushed into this repository with an image tag based on
+ * the source hash.
+ */
+const ASSETS_ECR_REPOSITORY_NAME = 'aws-cdk/assets';
+
+/**
+ * This allows users to work around the fact that the ECR repository is
+ * (currently) not configurable by setting this context key to their desired
+ * repository name. The CLI will auto-create this ECR repository if it's not
+ * already created.
+ */
+const ASSETS_ECR_REPOSITORY_NAME_OVERRIDE_CONTEXT_KEY = 'assets-ecr-repository-name';
+
+/**
+ * Encodes information how a certain Stack should be deployed
+ */
+export interface IDeploymentConfiguration {
+  /**
+   * Bind to the stack this environment is going to be used on
+   *
+   * Must be called before any of the other methods are called.
+   */
+  bind(stack: Stack): void;
+
+  /**
+   * Register a File Asset
+   *
+   * Returns the parameters that can be used to refer to the asset inside the template.
+   */
+  addFileAsset(asset: FileAssetSource): FileAssetLocation;
+
+  /**
+   * Register a Docker Image Asset
+   *
+   * Returns the parameters that can be used to refer to the asset inside the template.
+   */
+  addDockerImageAsset(asset: DockerImageAssetSource): DockerImageAssetLocation;
+
+  /**
+   * Synthesize all artifacts required for the stack into the session
+   *
+   * @experimental
+   */
+  writeStackArtifacts(session: ISynthesisSession): void;
```

Comment:
It's not intuitively clear to me what this is supposed to do? What is the use case of overriding this?

Reply:
The use case is, effectively the synthesis for stacks is different depending on the deployment provider. So we outsource the synthesis to an object, and the object becomes the point of polymorphism.